### PR TITLE
Add medians

### PR DESCRIPTION
**Pull request by qbarthelemy**
This PR:
- [X] adds geometric medians for Euclidean and Riemannian metrics;
- [X] completes tests;
- [x] adds an example on 2x2 SPD matrices, comparing means and medians for Euclidean and Riemannian metrics, but also the center of the Riemannian potato (robust mean by iterative outlier removal).

### CODE REVIEW

**Comment by qbarthelemy** [comment]
PR is ready.

I display plots of example comparing means and medians: (i) vector space, (ii) manifold of 2x2 SPD matrices, (iii) zoom of (ii).

Figure_1
Figure_2
Figure_3

File: examples/simulated/plot_mean_median_comparison.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+# License: BSD (3-clause)
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn.datasets import make_blobs
+from pyriemann.datasets import make_outliers
+from pyriemann.utils.mean import (
+    mean_euclid, mean_riemann, median_euclid, median_riemann
+)
```
Comment (agramfort) [inline examples/simulated/plot_mean_median_comparison.py:21]: we should be able to do:

```suggestion
from pyriemann.utils import (
    mean_euclid, mean_riemann, median_euclid, median_riemann
)
```

importing median from mean is weird and it would reduce the nesting of the code.

File: examples/simulated/plot_mean_median_comparison.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+###############################################################################
+# Photo finish
+# ------------
+#
+# Specific zoom on means and medians, with a surprise guest.
+#
+# Riemannian potato is fitted with an iterative outlier removal, providing a
+# robust mean [4]_.
+
+C_rp = Potato(metric='riemann', threshold=1.5).fit(X)._mdm.covmeans_[0]
```
Comment (agramfort) [inline examples/simulated/plot_mean_median_comparison.py:117]: you are exposing a private attribute in a public example

File: examples/simulated/plot_mean_median_comparison.py (hunk 1 above)
Comment (agramfort) [inline examples/simulated/plot_mean_median_comparison.py:21]: we should be able to do:

```suggestion
from pyriemann.utils import (
    mean_euclid, mean_riemann, median_euclid, median_riemann
)
```

importing median from mean is weird and it would reduce the nesting of the code.

File: pyriemann/utils/mean.py [hunk 3]
```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
         sample_weight=sample_weight
     )
     return C
+
+
+###############################################################################
+
+
+def median_euclid(covmats, tol=10e-6, maxiter=50, init=None):
```
Comment (agramfort) [inline pyriemann/utils/mean.py:742]: In would move this to median.py

File: pyriemann/utils/mean.py [hunk 4]
```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
...
+    """Median of matrices according to the Euclidean metric.
+
+    Return the Euclidean geometric median for matrices,
+    different from the marginal median provided by numpy.median().
```
Comment (agramfort) [inline pyriemann/utils/mean.py:746]: can you add the formula of what is computed in ..math mode?

File: pyriemann/utils/mean.py [hunk 5]
```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
...
+
+    Parameters
+    ----------
+    covmats : ndarray, shape (n_matrices, n_channels, n_channels)
+        Set of matrices.
+    tol : float, default=10e-6
+        The tolerance to stop the gradient descent.
```
Comment (agramfort) [inline pyriemann/utils/mean.py:753]: it's not really a gradient descent you do here don't you? to me it's more like iterative reweighted least square

File: pyriemann/utils/mean.py [hunk 6]
```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
...
+        dists = pairwise_distance(
+            covmats,
+            C[np.newaxis, ...],
+            metric='euclid'
+        )[:, 0]
+        is_zero = (dists == 0)
+
+        inv_dists = 1 / dists[~is_zero]
+        weights = inv_dists / np.sum(inv_dists)
+        Cnew = np.einsum('a,abc->bc', weights, covmats[~is_zero])
```
Comment (agramfort) [inline pyriemann/utils/mean.py:789]: how about rather adding weights param to mean_euclid and iterate calls to mean_euclid?

File: pyriemann/utils/mean.py [hunk 7]
```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
...
+    """Median of SPD matrices according to the Riemannian metric.
+
+    Return the affine-invariant Riemannian geometric median for SPD matrices.
+
+    Parameters
+    ----------
+    covmats : ndarray, shape (n_matrices, n_channels, n_channels)
+        Set of SPD matrices.
+    tol : float, default=10e-6
+        The tolerance to stop the gradient descent.
```
Comment (agramfort) [inline pyriemann/utils/mean.py:817]: idem

File: pyriemann/utils/mean.py [hunk 8]
```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
...
+    References
+    ----------
+    .. [1] Fletcher PT, Venkatasubramanian S and Joshi S. "The geometric median
+        on Riemannian manifolds with application to robust atlas estimation",
+        NeuroImage, 2009
+    .. [2] Yang L, Arnaudon M and Barbaresco F. "Riemannian median, geometry of
+        covariance matrices and radar target detection", EURAD, 2010
+    """
+    if init is None:
+        C = mean_euclid(covmats)
```
Comment (agramfort) [inline pyriemann/utils/mean.py:842]: why not mean_riemann?

Comment (qbarthelemy) [reply pyriemann/utils/mean.py:842]: I don’t know if it’s a safe idea to initialize an iterative algorithm using another iterative algorithm that may not converge.

File: pyriemann/utils/mean.py (hunk 5 above)
Comment (qbarthelemy) [reply pyriemann/utils/mean.py:753]: From Section 5 of [Fletcher2009 ](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735114/):
> The procedure iteratively updates the estimate mk of the geometric median using essentially a steepest descent on the weighted sum-of-distances function

File: pyriemann/utils/mean.py (hunk 4 above)
Comment (qbarthelemy) [reply pyriemann/utils/mean.py]: Great idea! it's done!

File: pyriemann/utils/mean.py (hunk 6 above)
Comment (qbarthelemy) [reply pyriemann/utils/mean.py]: Ok. I must read papers again.

File: pyriemann/utils/median.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        dists = pairwise_distance(
+            X,
+            M[np.newaxis, ...],
+            metric='euclid'
+        )[:, 0]
+        is_zero = (dists == 0)
+
+        inv_dists = 1 / dists[~is_zero]
+        weights = inv_dists / np.sum(inv_dists)
+        Mnew = np.einsum('a,abc->bc', weights, X[~is_zero])
```
Comment (agramfort) [inline pyriemann/utils/median.py:66]: to me this is not a gradient descent. A gradient descent would look like:

Mnew = Mold - step_size * something

where is the step size in your code?

File: pyriemann/utils/median.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        NeuroImage, 2009
+    .. [2] Yang L, Arnaudon M and Barbaresco F. "Riemannian median, geometry of
+        covariance matrices and radar target detection", EURAD, 2010
+    """
+    if init is None:
+        M = mean_euclid(X)
+    else:
+        M = init
+
+    nu = 1.0
```
Comment (agramfort) [inline pyriemann/utils/median.py:127]: that's your step size? can you call it step_size? It's the alpha in https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735114/pdf/nihms100397.pdf that should be between 0 and 2?

File: pyriemann/utils/median.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    tau = np.finfo(np.float64).max
+    crit = np.finfo(np.float64).max
```
Comment (agramfort) [inline pyriemann/utils/median.py:129]: can you use more explicit names for tau and crit?

File: doc/whatsnew.rst [hunk 12]
```diff
@@ -19,7 +19,9 @@ v0.3.1.dev
 - Enhance AJD: add ``init`` to :func:`pyriemann.utils.ajd.ajd_pham` and :func:`pyriemann.utils.ajd.rjd`,
   add ``warm_restart`` to :class:`pyriemann.spatialfilters.AJDC`. :pr:`196` by :user:`qbarthelemy`
 
-- Add parameter ``sampling_method`` to :func:`pyriemann.dataset.sample_gaussian_spd`, with ``rejection`` accelerating 2x2 matrices generation. :pr:`198` by :user:`Artim436`
+- Add parameter ``sampling_method`` to :func:`pyriemann.datasets.sample_gaussian_spd`, with ``rejection`` accelerating 2x2 matrices generation. :pr:`198` by :user:`Artim436`
+
+- Add geometric medians for Euclidean and Riemannian metrics: :func:`pyriemann.utils.median.median_euclid` and :func:`pyriemann.utils.median.median_riemann`. :pr:`200` by :user:`qbarthelemy`
```
Comment (agramfort) [inline doc/whatsnew.rst:24]: I am not a fan the repetition of median in the import name 

would you object to :

:func:`pyriemann.utils.median_riemann`

or you think that utils module is already too crowded?

File: examples/simulated/plot_mean_median_comparison.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+#
+# License: BSD (3-clause)
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn.datasets import make_blobs
+from pyriemann.datasets import make_outliers
+from pyriemann.utils.mean import mean_euclid, mean_riemann
+from pyriemann.utils.median import median_euclid, median_riemann
```
Comment (agramfort) [inline examples/simulated/plot_mean_median_comparison.py:20]:
```suggestion
from pyriemann.utils import mean_euclid, mean_riemann
from pyriemann.utils import median_euclid, median_riemann
```

???

File: doc/whatsnew.rst (hunk 12 above)
Comment (qbarthelemy) [reply doc/whatsnew.rst:24]: Do you suggest to add these imports in `utils.__init__` file?
```
from .mean import (
    mean_covariance,
    mean_ale,
    mean_alm,
    mean_euclid,
    mean_harmonic,
    mean_identity,
    mean_kullback_sym,
    mean_logdet,
    mean_logeuclid,
    mean_power,
    mean_riemann,
    mean_wasserstein,
    maskedmean_riemann,
    nanmean_riemann,
)
from .median import (
    median_euclid,
    median_riemann,
)
```

**Comment by agramfort** [comment]
Yes. What do you think ?

File: doc/whatsnew.rst (hunk 12 above)
Comment (qbarthelemy) [reply doc/whatsnew.rst:24]: > Yes. What do you think ?

Ok for me!

File: pyriemann/utils/median.py (hunk 9 above)
Comment (qbarthelemy) [reply pyriemann/utils/median.py:66]: You are right. Ostresh's implementation is a gradient descent, not Vardi's one.
I change the doc.

File: doc/api.rst [hunk 14]
```diff
@@ -221,6 +221,17 @@ Means
     maskedmean_riemann
     nanmean_riemann
 
+Medians
+~~~~~~~~~~~~~~~~~~~~~~
+.. _median_api:
+.. currentmodule:: pyriemann.utils.median
```
Comment (agramfort) [inline doc/api.rst:227]: I guess this can be now
```suggestion
.. currentmodule:: pyriemann.utils
```

**Review by agramfort — APPROVED** [review]
just a nitpick

🙏 @qbarthelemy

**Comment by agramfort** [comment]
Thx @qbarthelemy 🙏
